pkg/kubelet/images/pullmanager/mem_pullrecords.go

```diff
@@ -0,0 +1,186 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pullmanager
+
+import (
+	"strings"
+	"sync"
+
+	kubeletconfiginternal "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	"k8s.io/utils/lru"
+)
+
+type LRUCache[K comparable, V any] struct {
+	cache *lru.Cache
+}
+
+func NewLRUCache[K comparable, V any](size int) *LRUCache[K, V] {
+	return &LRUCache[K, V]{
+		cache: lru.New(size),
+	}
+}
+
+func (c *LRUCache[K, V]) Get(key K) (*V, bool) {
+	value, found := c.cache.Get(key)
+	if !found {
+		return nil, false
+	}
+	if value == nil {
+		return nil, true
+	}
+	return value.(*V), true
+}
+
+func (c *LRUCache[K, V]) Set(key K, value *V) { c.cache.Add(key, value) }
+func (c *LRUCache[K, V]) Delete(key K)        { c.cache.Remove(key) }
+func (c *LRUCache[K, V]) Len() int            { return c.cache.Len() }
+
+// cachedPullRecordsAccessor implements a write-through cache layer on top
+// of another PullRecordsAccessor
+type cachedPullRecordsAccessor struct {
+	delegate PullRecordsAccessor
+
+	intentsMutex       *sync.RWMutex
+	intents            *LRUCache[string, kubeletconfiginternal.ImagePullIntent]
+	pulledRecordsMutex *sync.RWMutex
+	pulledRecords      *LRUCache[string, kubeletconfiginternal.ImagePulledRecord]
+}
+
+func NewCachedPullRecordsAccessor(delegate PullRecordsAccessor) *cachedPullRecordsAccessor {
+	// TODO: handle errors from records listing here and below
+	coldIntents, _ := delegate.ListImagePullIntents()
+	warmIntents := NewLRUCache[string, kubeletconfiginternal.ImagePullIntent](50)
+	for _, intent := range coldIntents[:min(50, len(coldIntents))] {
+		warmIntents.Set(intent.Image, intent)
+	}
+
+	coldPulledRecords, _ := delegate.ListImagePulledRecords()
+	warmPulledRecords := NewLRUCache[string, kubeletconfiginternal.ImagePulledRecord](100)
+	for _, pulledRecord := range coldPulledRecords[:min(100, len(coldPulledRecords))] {
+		warmPulledRecords.Set(pulledRecord.ImageRef, pulledRecord)
+	}
+
+	return &cachedPullRecordsAccessor{
+		delegate: delegate,
+
+		intentsMutex:       &sync.RWMutex{},
+		intents:            warmIntents,
+		pulledRecordsMutex: &sync.RWMutex{},
+		pulledRecords:      warmPulledRecords,
+	}
+}
+
+func (c *cachedPullRecordsAccessor) ListImagePullIntents() ([]*kubeletconfiginternal.ImagePullIntent, error) {
+	//TODO: maybe this should be protected by a read lock
+	c.intentsMutex.RLock()
+	defer c.intentsMutex.RUnlock()
+	return c.delegate.ListImagePullIntents()
+}
+
+func (c *cachedPullRecordsAccessor) ImagePullIntentExists(image string) (bool, error) {
+	if _, exists := c.intents.Get(image); exists {
+		return true, nil
+	}
+
+	c.intentsMutex.RLock()
+	defer c.intentsMutex.RUnlock()
```

Comment:
it's... odd to check c.intents locally, then take a global **read** lock, then **write** to c.intents

what is this attempting to protect by locking?

Reply:
uploaded https://gist.github.com/liggitt/49f9eb909daea6f6e24c97ce73053591 with what we ended up with after in person code review